libphal/phal_pdbg.C

```diff
@@ -323,4 +326,121 @@ bool hasControlTransitionedToHost()
 	return (l_coreScratchRegData == 0);
 }
 
+std::optional<LocationCode> getUnexpandedLocCode(const LocationCode &locCode)
+{
+	// Location code should start with "U"
+	if (locCode[0] != 'U') {
+		log(level::ERROR,
+		    "Location code should start with \"U\""
+		    " but, given location code %s",
+		    locCode);
+		return std::nullopt;
+	}
+
+	// Given location code length should be need to match with minimum
+	// length
+	// to drop expanded format in given location code.
+	if (locCode.length() < EXP_LOCATIN_CODE_MIN_LENGTH) {
+		log(level::ERROR,
+		    "Given location code %s is not meet "
+		    "with minimum length %d",
+		    locCode, EXP_LOCATIN_CODE_MIN_LENGTH);
+		return std::nullopt;
+	}
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATIN_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode);
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATIN_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+int getAttrType(const LocationCode &locationCode, ATTR_TYPE_Type &fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	// Defining a list of existing frus
+	// Note: We need to update the list if a new fru is added
+	std::set<const char *> frus{"bmc", "proc", "tpm", "dimm"};
+	struct pdbg_target *target;
+
+	for (auto fru : frus) {
+		pdbg_for_each_class_target(fru, target)
+		{
+			if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode)) {
+				continue;
+			} else {
+				if (locCode == locationCode) {
+					if (DT_GET_PROP(ATTR_TYPE, target,
+							fruType)) {
+						return openpower::phal::
+						    exception::
+							DEVTREE_ATTR_READ_FAIL;
+					} else {
+						return 0; // success
+					}
+				}
+			}
+		}
+	}
+	return -1;
```

Comment:
It could be new error code to indicate the location code is not matching with predefined FRUs list so that its helps to fix the predefined list.

Reply:
a 0 or -1 is sufficient here as it is a small function rather than complicating with various return codes

or follow on how other methods are done in this file 

throw pdbgError_t(exception::PDBG_TARGET_INVALID);